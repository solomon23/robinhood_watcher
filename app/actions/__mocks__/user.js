const tryAuthFunction = jest.fn()

export function tryAuth() {
  return tryAuthFunction
}

export default {
  tryAuth,
}
