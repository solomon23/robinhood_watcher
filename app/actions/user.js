import { authenticate, tryAuthenticate } from '../services/api'
import type { Dispatch } from '../reducers/types'

export const LOGIN_USER_LOADING = 'LOGIN_USER_LOADING'
export const LOGIN_USER_ERROR = 'LOGIN_USER_ERROR'
export const LOGIN_USER_SUCCESS = 'LOGIN_USER_SUCCESS'
export const LOGIN_USER_NEEDS_MFA = 'LOGIN_USER_NEEDS_MFA'
export const USER_LOGOUT = 'USER_LOGOUT'
export const SET_USER_ACCOUNT_NUMBER = 'SET_USER_ACCOUNT_NUMBER'

export function login(username, password, mfaCode) {
  return (dispatch: Dispatch) => {
    dispatch({ type: LOGIN_USER_LOADING })

    return authenticate(username, password, mfaCode)
      .then(response => {
        if (response.twoFactorAuthRequired && !mfaCode) {
          dispatch({ type: LOGIN_USER_NEEDS_MFA })
        } else if (response.authenticated) {
          dispatch({ type: LOGIN_USER_SUCCESS, response })
        } else {
          dispatch({ type: LOGIN_USER_ERROR, message: response.error })
        }

        return null
      }).catch(err => {
        dispatch({ type: LOGIN_USER_ERROR, error: err })
      })
    }
}

export function tryAuth() {
  return (dispatch: Dispatch) => {
    dispatch({ type: LOGIN_USER_LOADING })

    return tryAuthenticate()
      .then(response => {
        if (response.authenticated) {
          dispatch({ type: LOGIN_USER_SUCCESS, response })
        } else {
          dispatch({ type: LOGIN_USER_ERROR, message: response.error })
        }

        return null
      }).catch(err => {
        dispatch({ type: LOGIN_USER_ERROR, error: err })
      })
    }
}

export function logout() {
  return {
    type: USER_LOGOUT,
  }
}