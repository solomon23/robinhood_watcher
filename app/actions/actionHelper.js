// @flow
import * as userActions from './user'
import * as appActions from './app'

export function createRequestTypes(base: string) {
  return ['REQUEST', 'SUCCESS', 'FAILURE'].reduce((acc, type) => {
    acc[type] = `${base}_${type}`
    return acc
  }, {})
}

export function action(type: string, payload: any = {}) {
  return { type, ...payload }
}

export function createRequests(base: any) {
  return {
    request: (o: any) => action(base.REQUEST, o),
    success: (o: any) => action(base.SUCCESS, o),
    failure: (o: any) => action(base.FAILURE, o),
  }
}

export function doApiCall(options: any) {
  const defaults = {
    api: () => { },
    action: {},
    apiParams: {},
    guardFunc: () => true,
    postFunc: (dispatch, getState, response) => response,
  }
  const opts = { ...defaults, ...options }
  opts.action = Array.isArray(options.action) ? options.action : [options.action]

  return (dispatch: Dispatch, getState: GetState) => {
    // bail if we don't pass any rules sent in
    if (!opts.guardFunc(getState())) {
      return Promise.resolve()
    }

    const { user } = getState()
    if (!user.authenticated) {
      // bail
      return Promise.resolve()
    }

    const params = { ...opts.apiParams || {}, user }

    // mark the call as starting
    opts.action.map(act => dispatch(act.request({ request: params })))

    // call the api with the args
    return opts.api(params)
      .then(
        response => opts.postFunc(dispatch, getState, response)
      ).then((response) => {
        opts.action.map(act => dispatch(act.success({ request: params, response })))

        return response
      }, (error) => {
        opts.action.map(act => dispatch(act.failure({ request: params, response: error })))
        return Promise.reject(error)
      }).catch((error) => {
        if (error === 401) {
          // log the user out and stop refreshing the data since our auth token is no longer valid
          dispatch(userActions.logout())
          dispatch(appActions.stopRefresh())
        }
      })
  }
}
