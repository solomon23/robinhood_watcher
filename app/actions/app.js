// @flow
import type { GetState, Dispatch } from '../reducers/types'
import * as API from '../services/api'
import { createRequests, doApiCall, createRequestTypes } from './actionHelper'

export const START_REFRESH_TIMER = 'START_REFRESH_TIMER'
export const STOP_REFRESH_TIMER = 'STOP_REFRESH_TIMER'

export const GET_ACCOUNT_NUMBER = createRequestTypes('GET_ACCOUNT_NUMBER')
export const getAccountNumberActions = createRequests(GET_ACCOUNT_NUMBER)

export const GET_PORTFOLIO = createRequestTypes('GET_PORTFOLIO')
export const getPortfolioActions = createRequests(GET_PORTFOLIO)

export const GET_ACCOUNT_POSITIONS = createRequestTypes('GET_ACCOUNT_POSITIONS')
export const getAccountPositionsActions = createRequests(GET_ACCOUNT_POSITIONS)

export function startRefreshTimer() {
  return { type: START_REFRESH_TIMER }
}

export function stopRefresh() {
  return { type: STOP_REFRESH_TIMER }
}

export function getAccount() {
  return doApiCall({
    api: API.getAccountNumber,
    action: getAccountNumberActions,
    guardFunc: (state) => {
      const { user } = state
      return !user.accountNumber
    },
  })
}

export function getPortfolio() {
  return doApiCall({
    api: API.getPortfolio,
    action: getPortfolioActions,
  })
}

export function getAccountPositions() {
  return doApiCall({
    api: API.getAccountPositions,
    action: getAccountPositionsActions,
  })
}

export function getAllData() {
  return async function getAllCallback(dispatch: Dispatch, getState: GetState) {
    // (async () => {
      const { user } = getState()

      // bail if the user isn't logged in
      if (!user.authenticated) {
        return
      }

      if (!user.accountNumber) {
        await getAccount()(dispatch, getState)
      }

      // const positions = await API.getAccountPositions(user)
      await getPortfolio()(dispatch, getState)
      await getAccountPositions()(dispatch, getState)
    // })()
  }
}
