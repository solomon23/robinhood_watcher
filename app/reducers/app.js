// @flow
import { GET_PORTFOLIO, GET_ACCOUNT_POSITIONS, STOP_REFRESH_TIMER, START_REFRESH_TIMER } from '../actions/app'
import type { Action } from './types'

const defaultState = {
  actions: {
    portfolioLoading: false,
    positionsLoading: false,
  },
  portfolio: {
  },
  positions: [
  ],
  refreshing: false,
}

export default function app(state = defaultState, action: Action) {
  switch (action.type) {
    case START_REFRESH_TIMER:
      return { ...state, refreshing: true }

    case STOP_REFRESH_TIMER:
      return { ...state, refreshing: false }

    case GET_PORTFOLIO.REQUEST: {
      const actions = { ...state.actions, portfolioLoading: true }
      return { ...state, actions }
    }

    case GET_PORTFOLIO.FAILURE: {
      const actions = { ...state.actions, portfolioLoading: false }
      return { ...state, actions }
    }

    case GET_PORTFOLIO.SUCCESS: {
      const actions = { ...state.actions, portfolioLoading: false }
      return { ...state, actions, portfolio: action.response }
    }

    case GET_ACCOUNT_POSITIONS.REQUEST: {
      const actions = { ...state.actions, positionsLoading: true }
      return { ...state, actions }
    }

    case GET_ACCOUNT_POSITIONS.FAILURE: {
      const actions = { ...state.actions, positionsLoading: false }
      return { ...state, actions }
    }

    case GET_ACCOUNT_POSITIONS.SUCCESS: {
      const actions = { ...state.actions, positionsLoading: false }
      return { ...state, actions, positions: action.response }
    }

    default:
      return state
  }
}
