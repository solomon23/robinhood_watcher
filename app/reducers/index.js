// @flow
import { combineReducers } from 'redux'
import { connectRouter } from 'connected-react-router'
import toolbar from './toolbar'
import user from './user'
import app from './app'

export default function createRootReducer(history: {}) {
  const routerReducer = connectRouter(history)(() => {})

  return connectRouter(history)(
    combineReducers({
      router: routerReducer,
      toolbar,
      app,
      user,
    })
  )
}
