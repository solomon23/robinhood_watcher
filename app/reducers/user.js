// @flow
import * as userActions from '../actions/user'
import * as appActions from '../actions/app'
import type { Action } from './types'

const defaultState = {
  isLoading: false,
  isError: false,
  errorReason: null,
  needsMfa: false,
  authenticated: false,
  accountNumber: null,
}

export default function app(state = defaultState, action: Action) {
  switch (action.type) {
    case userActions.LOGIN_USER_LOADING:
      return { ...defaultState, isLoading: true, authenticated: false }

    case userActions.LOGIN_USER_ERROR:
      return { ...defaultState, isError: true, errorReason: action.message }

    case userActions.LOGIN_USER_NEEDS_MFA:
      return { ...defaultState, needsMfa: true, authenticated: false }

    case userActions.LOGIN_USER_SUCCESS:
      return { ...defaultState, authenticated: action.response.authenticated }

    case appActions.GET_ACCOUNT_NUMBER.SUCCESS:
      return { ...state, accountNumber: action.response }

    case userActions.USER_LOGOUT:
      return { ...defaultState, authenticated: false }

    default:
      return state
  }
}
