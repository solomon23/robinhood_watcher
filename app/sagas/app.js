import { takeEvery, fork, put, call, race, take } from 'redux-saga/effects'
import * as appActions from '../actions/app'

const REFRESH_TIME = 1000 * 60 * 10

const wait = ms => (
  new Promise(resolve => {
    setTimeout(() => resolve(), ms)
  })
)

function* handleStartRefresh() {
  yield takeEvery(appActions.START_REFRESH_TIMER, function* handler() {
    // fire it off first
    yield put(appActions.getAllData())

    while(true) {
      const timer = yield race({
        stopped: take(appActions.STOP_REFRESH_TIMER),
        tick: call(wait, REFRESH_TIME),
      })

      if (!timer.stopped) {
        yield put(appActions.getAllData())
      } else {
        break
      }
    }
  })
}

function* app() {
  yield fork(handleStartRefresh)
}

export default app
