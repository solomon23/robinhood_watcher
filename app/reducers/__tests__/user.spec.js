import reducer from '../user'
import * as appActions from '../../actions/app'
import * as userActions from '../../actions/user'

const initialState = {
  isLoading: false,
  isError: false,
  errorReason: null,
  needsMfa: false,
  authenticated: false,
  accountNumber: null,
}

describe('User reducer', () => {
  it('should return the initial state', () => {
    expect(
      reducer(undefined, {})
    ).toEqual(initialState)
  })

  it('should handle LOGIN_USER_LOADING', () => {
    expect(
      reducer({}, {
        type: userActions.LOGIN_USER_LOADING,
      })
    ).toMatchSnapshot()
  })

  it('should handle LOGIN_USER_ERROR', () => {
    expect(
      reducer({}, {
        type: userActions.LOGIN_USER_ERROR,
        message: 'bad',
      })
    ).toMatchSnapshot()
  })

  it('should handle LOGIN_USER_NEEDS_MFA', () => {
    expect(
      reducer({}, {
        type: userActions.LOGIN_USER_NEEDS_MFA,
      })
    ).toMatchSnapshot()
  })

  it('should handle LOGIN_USER_SUCCESS', () => {
    expect(
      reducer({}, {
        type: userActions.LOGIN_USER_SUCCESS,
        response: { authenticated: true },
      })
    ).toMatchSnapshot()
  })

  it('should handle GET_ACCOUNT_NUMBER.SUCCESS', () => {
    expect(
      reducer({}, {
        type: appActions.GET_ACCOUNT_NUMBER.SUCCESS,
        response: '123',
      })
    ).toMatchSnapshot()
  })

  it('should handle USER_LOGOUT', () => {
    expect(
      reducer({ authenticated: true }, {
        type: userActions.USER_LOGOUT,
      })
    ).toMatchSnapshot()
  })
})
