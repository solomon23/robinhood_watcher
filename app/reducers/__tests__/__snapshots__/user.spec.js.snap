// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User reducer should handle GET_ACCOUNT_NUMBER.SUCCESS 1`] = `
Object {
  "accountNumber": "123",
}
`;

exports[`User reducer should handle LOGIN_USER_ERROR 1`] = `
Object {
  "accountNumber": null,
  "authenticated": false,
  "errorReason": "bad",
  "isError": true,
  "isLoading": false,
  "needsMfa": false,
}
`;

exports[`User reducer should handle LOGIN_USER_LOADING 1`] = `
Object {
  "accountNumber": null,
  "authenticated": false,
  "errorReason": null,
  "isError": false,
  "isLoading": true,
  "needsMfa": false,
}
`;

exports[`User reducer should handle LOGIN_USER_NEEDS_MFA 1`] = `
Object {
  "accountNumber": null,
  "authenticated": false,
  "errorReason": null,
  "isError": false,
  "isLoading": false,
  "needsMfa": true,
}
`;

exports[`User reducer should handle LOGIN_USER_SUCCESS 1`] = `
Object {
  "accountNumber": null,
  "authenticated": true,
  "errorReason": null,
  "isError": false,
  "isLoading": false,
  "needsMfa": false,
}
`;

exports[`User reducer should handle USER_LOGOUT 1`] = `
Object {
  "accountNumber": null,
  "authenticated": false,
  "errorReason": null,
  "isError": false,
  "isLoading": false,
  "needsMfa": false,
}
`;
